### PR TITLE
Hts 324: Splun events

### PR DESCRIPTION
**Pull request by adwelly**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala [hunk 1]
```diff
@@ -16,6 +16,8 @@
 
 package uk.gov.hmrc.helptosavefrontend.config
 
+import javax.inject.Inject
+
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala]: why is this import required ?

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 2]
```diff
@@ -67,6 +69,10 @@ class NSIConnectorImpl @Inject()(conf: Configuration) extends NSIConnector {
       .map { response ⇒
         response.status match {
           case Status.CREATED ⇒
+            val applicationEventResult = auditConnector.sendEvent(new ApplicationSubmittedEvent(userInfo))
+            applicationEventResult.onFailure {
+              case e: Throwable => Logger.error(s"Unable to post application submission event to audit connector - ${e.getMessage}", e)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: I doubt if the `e` at the end logs full stack trace ?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 3]
```diff
@@ -83,11 +87,32 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
+                        Ok(views.html.register.confirm_details(info))
+                      }
+                      case _ ⇒ {
+                        val eligibilityCheckEventResult = auditConnector.sendEvent(new EligibilityCheckEvent(nino, Some("Unknown eligibility problem")))
+                        eligibilityCheckEventResult.onFailure {
+                          case e: Throwable => Logger.error(s"Unable to post eligibility check event to audit connector - ${e.getMessage}", e)
+                        }
+
+                        SeeOther(routes.RegisterController.notEligible().url)
+                      }
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: i guess , instead of adding audit logic  in each connector , its better to have an Audit Object with sendEvent() function
and just call the method from each controllers.. so error handling goes out of each connector

File: app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        "address2" -> userInfo.contactDetails.address2,
+        "address3" -> {userInfo.contactDetails.address3.fold("") {identity}},
+        "address4" -> {userInfo.contactDetails.address4.fold("") {identity}},
+        "address5" -> {userInfo.contactDetails.address5.fold("") {identity}},
+        "postcode" -> userInfo.contactDetails.postcode,
+        "countryCode" -> {userInfo.contactDetails.countryCode.fold("") {identity}},
+        "email" -> userInfo.contactDetails.email,
+        "phoneNumber" -> {userInfo.contactDetails.phoneNumber.fold(""){identity}},
+        "communicationPreference" -> userInfo.contactDetails.communicationPreference,
+        "registrationChannel" -> userInfo.registrationChannel))
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala]: why do we want to log each and every user info field, is just nino not enough ?

File: app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala (hunk 1 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala]: Good spot - I will remove it

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala (hunk 2 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: It just logs the message. One of the rules in MDTP best practice for logging is never, ever log a stack trace.

File: app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala (hunk 4 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala]: This is what the transaction monitoring team requested.

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala (hunk 3 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: I'm not convinced that additional abstraction here helps readability, it follows a standard simple pattern that seems fairly straightforward. Happy to discuss.

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 5]
```diff
@@ -19,14 +19,16 @@ package uk.gov.hmrc.helptosavefrontend.connectors
...
 import uk.gov.hmrc.helptosavefrontend.config.FrontendAppConfig.{nsiAuthHeaderKey, nsiBasicAuth, nsiUrl}
-import uk.gov.hmrc.helptosavefrontend.config.WSHttpProxy
+import uk.gov.hmrc.helptosavefrontend.config.{FrontendAuditConnector, WSHttpProxy}
 import uk.gov.hmrc.helptosavefrontend.connectors.NSIConnector.{SubmissionFailure, SubmissionResult, SubmissionSuccess}
-import uk.gov.hmrc.helptosavefrontend.models.NSIUserInfo
+import uk.gov.hmrc.helptosavefrontend.models.{ApplicationSubmittedEvent, NSIUserInfo}
+import uk.gov.hmrc.helptosavefrontend.util.HTSAuditor
 import uk.gov.hmrc.helptosavefrontend.util.HttpResponseOps._
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
 import uk.gov.hmrc.play.http._
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: i think we are not using `Application` ,  `FrontendAuditConnector`   and `uk.gov.hmrc.play.audit.http.connector.AuditConnector`  anymore in this class?

File: app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala (hunk 4 above)
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala]: ok, i thought we are not supposed to log customer's private information like dob.. etc..

File: app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+package uk.gov.hmrc.helptosavefrontend.models
+
+import uk.gov.hmrc.play.audit.model.DataEvent
+import uk.gov.hmrc.play.http.HeaderCarrier
+import uk.gov.hmrc.play.audit.AuditExtensions._
+
+abstract class HTSEvent(auditType: String, detail: Map[String,String])(implicit hc: HeaderCarrier)
+  extends DataEvent(auditSource = "hts-frontend", auditType = auditType, detail = detail, tags = hc.toAuditTags("", "N/A"))
+
+class ApplicationSubmittedEvent(userInfo: NSIUserInfo)(implicit hc: HeaderCarrier)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/models/HTSEvent.scala]: I would go for `AccountCreationEvent`
